### DISCUSSION TITLE
PHcontrol+e

### DISCUSSION THREAD

**Discussion by nielsvanbaal** [davidm-glitch/home-assistant-eheim-digital#5] [Q&A]
Hey

I have an eheim PHcontrol+e set. And been trying to get the pH listed in homeassistant. I tried with scrape but so far no luck there too. In my search I came across this project.

My knowledge on this subject is not the best, but I would like to try.

I was able to find out the pH is being send by using firefox, page inspector, network, responses
And I was able to connect with a websocket client on my pc. But from there Im stuck in how to apply this to home assistant.

**Comment by davidm-glitch** [comment]
Hi @nielsvanbaal !

Happy to hear that there are even more people looking for these functionalities :)

Your way was correct in checking the websockets connection. You can help us by providing the communication with your device. 

When you take a look at the Wiki in this repository you can see what we were able to find out. 

We can add the ph control device inside the python file for now. We are still working on the integration which can be installed via hacs later.


EDIT: I also planned to buy the PHcontrol later. Are you happy with it?

**Reply by nielsvanbaal** [reply]
I previously build my own phcontroler. With an ESP (i believe the eheim one also is build around an ESP chip), but I couln't get the values smoothed out. They where all over the place, think this had something to do with the power supply. The eheim works a lot better. And I also have more faith in the eheim thing running at all times then my homeassistant / zigbee network. 

Only thing I think is a bit weird, I can't set it up to go lower then 6.8 unless I chose a absurd kH. But far as I have seen pH is more stable now and plants assimilate better. Think most fish will also appreciate the more steady pH.

What can I provide to help or what can I do myself?

Is this of use: 
{
  "nReduce" : -7,
  "acclimatization" : 0,
  "schedule" : [
    [
      450,
      70
    ],
    [
      1380,
      74
    ]
  ],
  "from" : "E0:5A:1B:61:88:18",
  "nightStartT" : 1320,
  "kH" : 6,
  "active" : 1,
  "partnerName" : "",
  "sollPH" : 68,
  "mode" : 1,
  "valveIsActive" : 0,
  "dayStartT" : 450,
  "hystHigh" : 1,
  "alertState" : 0,
  "offset" : 0,
  "hystLow" : 1,
  "isPH" : 68,
  "title" : "PH_DATA",
  "expert" : 0,
  "sync" : "",
  "serviceTime" : 58
}

This one only comes around once when I first connect with the websocket. Then its only from user to mac and keep alives.

My main wish is to have the isPH, so I can log that one. But being able to see and log when the valve is open or closed is also a nice to have.

**Comment by davidm-glitch** [comment]
Yes that is exactly what I need!

You could also try to send commands (change settings) from the webinterface and observe the data that is sent to the device. 

The thing with the setting of PH value is right in my opinion. As you can see in the attached table there is a correlation between PH, KH and CO2. You have set a KH value of 6 and with that the lowest possible PH you can reach is 6.8. If you would want to go lower you would have more than 30mg/l CO2 and that would be harmful for your animals!! Eheim is protecting you and your fish here.

![CO2_KH-PH-Tabelle_1600px](https://github.com/davidm-glitch/home-assistant-eheim-digital/assets/59320226/eed7d3f8-63a3-421b-95d4-8c0fdf9f02f4)

If you want lower PH you have to lower your KH ;-) 

I will start putting your data to the wiki and change the python file. 

Can you check for the commands? Maybe there is also a command for requesting the data that is sent only once in the beginning.

I would bet that it is something like this

`
{
    "title": "GET_PH_DATA",
    "to": "<MAC_ADDRESS>",
    "from": "USER"
}
`

You could try this with a tool like "Postman" or observe the websocket data when you click around in the webinterface.


EDIT: I updated the python file and the config inside the readme. Maybe you can try if it works?

**Comment by nielsvanbaal** [comment]
Yes thats the problem, I only have a " just working" understanding of a lot of things ;) 

I was able to get a response with sending a command over websocket, and I get data back.

I am running MQTT for my zigbee stuff, so that works.

Only step 3 I can't figure out.
Ive added pythonscripts to config yaml
and put the adjusted .py file in a folder 'python_scripts'  in my config

I get a ton of entities in my HA now. But no values in them. But I haven't broken anything on the plus side, and usually when I try something new in HA is will break atleast a few other things before working :)

Also looked up the sw versions, both website and server are 2.01.6

**Comment by davidm-glitch** [comment]
Okay very nice! Could you figure out the commands you sent?

Step 3 is a little tricky and I dont really know how to do this correctly. Sadly it wont work with pyscript because of the dependencies. 

I also tried some other things to get this running on my HA machine but no luck. Its running on my PC for now, but if you find a better solution feel free to share is. As long as the real integration is not finished ;)

**Comment by nielsvanbaal** [comment]
Yes the command you told me to works, and I get back the data just like on first connection. Turns out if I look long enough that data will also come around eventually. And when I refresh the page of the controller.

Far as I could figure out you can run python scripts with dependency when you make it into an addon. But no clue how. 

I also looked at the add on AppDaemon
It supports selecting of python packages to load.

In its config I put " pip install websockets" according to the log that works. But now Im clueless on how to run the py file.

**Comment by davidm-glitch** [comment]
I found a solution but I dont know how stable it is and I dont think it survives a restart but if you want you can check it out. It is updated in the readme file.

**Reply by nielsvanbaal** [reply]
It works!

I tried but doesn't survive a restart. But is working for now. 

Would be cool if the datatype (i think) would support a history graph. And maybe unique id's in the future.

**Comment by davidm-glitch** [comment]
@nielsvanbaal can you tell me the min and max values you can set for pH and KH? Then I could make the settings to be controllable via HA as well ;)

**Reply by nielsvanbaal** [reply]
Kh is 1 to 20 just like the table above.
Lowest ph wil go (with kh set to 1) is 6. And highest is 9.

**Comment by nielsvanbaal** [comment]
I tried the new way with hacs custom integration. But doesnt seem to work yet, is that right/known?

I can set it up and enter my IP for my phcontrol, but all that happens is this in the log:
Setup failed for filter_pump: Integration not found.
13:17:51 – (ERROR) setup.py
No request messages found for device type: filter_pump
13:17:51 – (WARNING) EHEIM Digital (custom integration) - message first occurred at 13:17:51 and shows up 3 times

there is one entry on the eheim digital integration, but it says no devices or entities. Is there a config I have to edit maybe?

**Comment by davidm-glitch** [comment]
That is exactly what should happen by now. Its still work in progress. We managed to make an integration that you can install and enter the IP. With that IP it scans for supported devices which is for testing now just the led control. But we are working hard on getting things done and understand how integrations in HA should work ;)

**Comment by nielsvanbaal** [comment]
Thnx!

How good are those leds? I want to change my regular (juwel stock led in a rio240). Tempted to go for the juwel helialux spectrum with wifi control etc or a eheim one.

**Reply by davidm-glitch** [reply]
I am quite happy with them, the brightness is quite good and the functions of the led control are also very nice. Can't say anything bad about it. The light bars were included in my tank, one "daylight" and one "plants".

Can't say anything about the Juwel one. I think it's about the same as the eheim ones. Might fit better to your setup? But I don't think that there will be a Juwel integration for HA in the future 😜

**Reply by nielsvanbaal** [reply]
https://community.home-assistant.io/t/custom-component-juwel-helialux-smart-controller/385515

😜

**Reply by davidm-glitch** [reply]
Uhh, nice! 🤩

Well then, it's up to you 😂

**Comment by jkhalil** [comment]
@nielsvanbaal - Would you be willing to download [Postman](https://www.postman.com/), connect to the EHEIM, and share the data that is received?  I can walk you through the process of how to get the data we need to add the PH Controller into our integration.

**Reply by nielsvanbaal** [reply]
Shure. If it runs in a mac M1 it shouldn't be a problem. I currently don't own a working windows pc. Atleast not one that's all put together. But tomorrow is my day off work so I might be able to throw some parts together and install windows on it if it's not possible on a mac.

**Comment by jkhalil** [comment]
It does indeed run on Mac, that's what I use.

Once you download and install it, we want to create a new websocket connection:
<img width="2456" alt="Screenshot 2023-08-17 at 12 46 19 PM" src="https://github.com/davidm-glitch/home-assistant-eheim-digital/assets/3517333/8350f576-806d-4908-b2b3-5c67fdb53725">

Then we want to put **ws://IP-ADDRESS/ws** with the ip address you used to connect to the web frontend 

<img width="2623" alt="Screenshot 2023-08-17 at 12 48 37 PM" src="https://github.com/davidm-glitch/home-assistant-eheim-digital/assets/3517333/f33902ec-a69e-4b06-aa63-a817e42683c1">


Once you hit connect on the right side, you will see some data stream in. 

Then I want you to go to the web frontend and just load the default page. You should start to see more data come strolling in like this:

<img width="1997" alt="Screenshot 2023-08-17 at 12 52 12 PM" src="https://github.com/davidm-glitch/home-assistant-eheim-digital/assets/3517333/536f45d0-7c5d-467b-a4cc-06e29ec07ae4">


We are looking for any messages that might look like this:

**{ "title": "GET_PH_DATA", "to": "E0:5A:1B:61:88:18", "from": "USER" }** 

And a response that would be right behind it:

 **{ "title": "PH_DATA", "from":"E0:5A:1B:61:88:18", "nightStartT" : 1320, "kH" : 6, "active" : 1, "partnerName" : "", "sollPH" : 68, "mode" : 1, "valveIsActive" : 0, "dayStartT" : 450, "hystHigh" : 1, "alertState" : 0, "offset" : 0, "hystLow" : 1, "isPH" : 68, "expert" : 0, "sync" : "", "serviceTime" : 58}** 

 Both of these would be very helpful.

Also, keep an eye out for anything that might have PH in the title section in any messages you see. You can also click on the device from the web page and see what calls are also being made to the device and the response for the data.

I hope that all of this doesn't confuse you.

let me know if you have anymore questions.

**Reply by nielsvanbaal** [reply]
Im getting 4 things on repeat
1:
{
    "title": "GET_PH_DATA",
    "to": "E0:5A:1B:61:88:18",
    "from": "USER"
}

2:
{
    "title": "GET_MESH_NETWORK",
    "to": "MASTER",
    "from": "USER"
}

3:
{
    "title": "USRDTA",
    "from": "E0:5A:1B:61:88:18",
    "name": "pH",
    "aqName": "Aquarium",
    "version": 9,
    "language": "EN",
    "timezone": 60,
    "tID": 30,
    "dst": 1,
    "tankconfig": "PH",
    "power": "9",
    "netmode": "ST",
    "host": "eheimdigital",
    "groupID": 0,
    "meshing": 1,
    "firstStart": 0,
    "revision": [
        2016,
        2016
    ],
    "latestAvailableRevision": [
        1008,
        1013,
        2016,
        2016
    ],
    "firmwareAvailable": 0,
    "emailAddr": "",
    "liveTime": 34140,
    "usrName": "",
    "unit": 0,
    "demoUse": 0,
    "sysLED": 100,
    "to": "USER"
}

4:
{
    "title": "PH_DATA",
    "from": "E0:5A:1B:61:88:18",
    "sollPH": 68,
    "isPH": 67,
    "active": 1,
    "hystLow": 1,
    "hystHigh": 1,
    "offset": 0,
    "valveIsActive": 0,
    "acclimatization": 0,
    "mode": 1,
    "expert": 0,
    "sync": "",
    "partnerName": "",
    "dayStartT": 450,
    "nightStartT": 1320,
    "nReduce": -6,
    "alertState": 0,
    "serviceTime": 36,
    "kH": 7,
    "schedule": [
        [
            450,
            82
        ],
        [
            1380,
            74
        ]
    ]
}

And when I change something:
{
    "title": "SET_PH_PARAM",
    "to": "E0:5A:1B:61:88:18",
    "sollPH": 68,
    "active": 1,
    "hystLow": 1,
    "hystHigh": 1,
    "offset": 0,
    "acclimatization": 0,
    "mode": 0,
    "expert": 0,
    "kH": 7,
    "schedule": [
        [
            450,
            82
        ],
        [
            1380,
            74
        ]
    ],
    "sync": "",
    "partnerName": "",
    "nightStartT": 1320,
    "dayStartT": 450,
    "nReduce": -6,
    "from": "USER"
}

Let me know if you need more / specific things!

**Comment by davidm-glitch** [comment]
@nielsvanbaal that looks good to me, can you explain what the time values and nReduce (i think night reduce) mean in the context of the ph control? Are these the times for turning off the co2 at night times? If so, what does the night reduce mean?
What can you do with acclimatization? And do you have any idea what the mode could be? Are there operating modes you can choose?

**Comment by nielsvanbaal** [comment]
There are 4 modes. 
Manual = just set one target pH. 
Bio simple = set 2 targets, one for day and other for night. You can define the times. I run this mode.
Bio expert = you can add more periods and set their goals (and apart from factory reset I havent found a way to delete extra entrys)
Smart mode = pairing with other EHEIM device I dont have any so I can't test what this does, but it gives me the choice between 3 led controller types and one filter.

I didn't find anything in the interface having to do with acclimatization. If you want I can arrange a portforward for you and whitelist your ip so you can look around yourself. 

Also found out that entering an e-mail results in an e-mail every night when the target switches, because then the difference between target and actual is to high. There is no way to chose what you want to have e-mailed. I would love a notification when I need to calibrate but I dont need an e-mail everytime the mode changes. The pH controller works fine but they could have done more in the software department. But then again its running on an ESP chip so not that much power behind it.

**Comment by davidm-glitch** [comment]
Thanks for you support!

I think that the smart mode is just synchronisation of the day and night times from the bio mode of the filter and / or for the led control. 
Is there some indicator in the webinterface that shows how long until the next calibration? I dont see a value in the websocket response that points to the calibration. 

We will come back to your offer with forwarding of the port! Thats the way @jkhalil is currently developing with my devices ;)

EDIT: Can you check which mode is what mode number in the websocket response?

**Comment by nielsvanbaal** [comment]
The days until calibration is serviceTime

Manual is mode 0
Bio is mode 1
Then expert is 0 or 1 depending on what you choose
Cant set mode to smart but judging by the simplicity I would guess that mode would be mode 2

**Reply by davidm-glitch** [reply]
Ah I didn't see the service time. Makes sense, as it is the same for filters.

I think @jkhalil can at least include the service time easily very fast ;)

**Reply by nielsvanbaal** [reply]
No rush, it still has 30 days left :)

**Comment by jkhalil** [comment]
@nielsvanbaal I just published release [v0.1.6](https://github.com/davidm-glitch/home-assistant-eheim-digital/releases/tag/v0.1.6) that has sensors for the PH Control. Please give it a try and let us know if the device shows up in the device register with the sensors attached.

Currently, we have not implemented any functionality to change any settings but it's on the roadmap.

**Comment by nielsvanbaal** [comment]
I noticed with v0.1.5 that it worked yesterday.
Upgraded, needed to add the device again. Now I have more sensors on the device.

![image](https://github.com/davidm-glitch/home-assistant-eheim-digital/assets/101596552/19a437cd-9a0b-4fe7-97a2-a20cd6125055)

But not the actual pH at the moment.
